### ISSUE TITLE
Dropping a cluster does not cancel active peeks

### ISSUE THREAD

**Issue by teskje** [MaterializeInc/materialize#16194]
### What version of Materialize are you using?

main

### How did you install Materialize?

Built from source

### What is the issue?

Peeks pending on a cluster are not canceled when the cluster is dropped.

This can be reproduced by issuing a `SELECT` with an `AS OF` in the future (requires `--unsafe-mode`):
```sql
SELECT * FROM mz_views AS OF 10000000000000000000;
```

Then open a second psql session and drop the cluster you ran the `SELECT` on:
```sql
DROP CLUSTER default;
```

Note how the `SELECT` in the first session is not canceled. If you manually cancel the `SELECT` (with ctrl-c), environmentd crashes:

```
thread 'coordinator' panicked at 'called `Result::unwrap()` on an `Err` value: InstanceMissing(User(1))', /Users/jan/devel/materialize/src/adapter/src/coord/peek.rs:478:22
stack backtrace:
   0: rust_begin_unwind
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
   2: core::result::unwrap_failed
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/result.rs:1785:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/result.rs:1107:23
   4: mz_adapter::coord::peek::<impl mz_adapter::coord::Coordinator<S>>::cancel_pending_peeks
             at ./src/adapter/src/coord/peek.rs:475:17
   5: mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator<S>>::handle_cancel
             at ./src/adapter/src/coord/command_handler.rs:528:18
   6: mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator<S>>::handle_command::{{closure}}
             at ./src/adapter/src/coord/command_handler.rs:93:17
   7: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
   8: mz_adapter::coord::message_handler::<impl mz_adapter::coord::Coordinator<S>>::message_command::{{closure}}::{{closure}}
             at ./src/adapter/src/coord/message_handler.rs:149:33
   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
  10: mz_adapter::coord::message_handler::<impl mz_adapter::coord::Coordinator<S>>::message_command::{{closure}}
             at ./src/adapter/src/coord/message_handler.rs:146:5
  11: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
  12: mz_adapter::coord::message_handler::<impl mz_adapter::coord::Coordinator<S>>::handle_message::{{closure}}
             at ./src/adapter/src/coord/message_handler.rs:40:63
  13: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
  14: mz_adapter::coord::Coordinator<S>::serve::{{closure}}
             at ./src/adapter/src/coord.rs:1001:37
  15: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/future/mod.rs:91:19
  16: tokio::runtime::park::CachedParkThread::block_on::{{closure}}
             at /Users/jan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/park.rs:272:63
  17: tokio::runtime::coop::with_budget
             at /Users/jan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/coop.rs:102:5
  18: tokio::runtime::coop::budget
             at /Users/jan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/coop.rs:68:5
  19: tokio::runtime::park::CachedParkThread::block_on
             at /Users/jan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/park.rs:272:31
  20: tokio::runtime::context::BlockingRegionGuard::block_on
             at /Users/jan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/context.rs:255:13
  21: tokio::runtime::handle::Handle::block_on
             at /Users/jan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/handle.rs:264:9
  22: mz_adapter::coord::serve::{{closure}}::{{closure}}
             at ./src/adapter/src/coord.rs:1147:17
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by teskje** [comment]
Related to https://github.com/MaterializeInc/materialize/issues/16121. Adapter should probably cancel any pending peeks when a cluster is dropped.

**Comment by teskje** [comment]
Duplicate of #15179.
